test/UniswapV2PairTest.sol

```diff
@@ -104,38 +140,84 @@ contract UniswapV2PairTest is Test {
         pair.swap(0, expected, address(this), '');
     }
 
+    function computeExpectedSwapAmount(uint swapAmount, uint token0Amount, uint token1Amount) private returns (uint256) {
+        // Compute new balance for token 0 with fees reduced (multiplied by 1000):
+        uint newBalance0Adjusted = ((token0Amount + swapAmount) * 1000) - (swapAmount * 3);
+
+        // Determine new target K value when the adjusted balances are multiplied (1000x what it should be to match previous units):
+        uint kAdjusted = token0Amount * token1Amount * 1000;
+
+        // Find the new target balance for token 1 and subtract it from the original amount to get expected output:
+        return token1Amount - divCeil(kAdjusted, newBalance0Adjusted);
+    }
+
     function testOptimisticTestCase1() public {
-        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(10), expandTo18Decimals(1));
+        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(10));
     }
 
     function testOptimisticTestCase2() public {
-        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(10), expandTo18Decimals(5), expandTo18Decimals(1));
+        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(10), expandTo18Decimals(5));
     }
 
     function testOptimisticTestCase3() public {
-        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(5), expandTo18Decimals(1));
+        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(5));
     }
 
     function testOptimisticTestCase4() public {
-        runOptimisticSwapTestCase(expandTo18Decimals(1), expandTo18Decimals(5), expandTo18Decimals(5), 1003009027081243732);
+        runOptimisticSwapTestCase(expandTo18Decimals(1), expandTo18Decimals(5), expandTo18Decimals(5));
+    }
+
+    function testOptimisticFuzzTestCase(uint outputVal, uint token0Val, uint token1Val) public {
+        (uint outputAmount, uint token0Amount, uint token1Amount) = generateOptimisticFuzzCase(outputVal, token0Val, token1Val);
+        runOptimisticSwapTestCase(outputAmount, token0Amount, token1Amount);
+    }
+
+    function generateOptimisticFuzzCase(uint outputAmountSeed, uint token0AmountSeed, uint token1AmountSeed) private
+            returns (uint _outputAmount, uint _token0Amount, uint _token1Amount) {
+
+        // Token 0 needs at least: A) one output amount to take out and B) one remaining, otherwise all removals will fail:
+        uint minToken0Amount = 2;
+        // Token 0 needs to leave at least 2 in the global supply for an input transfer that will be one or greater with fees (beyond one) taken out:
+        uint maxToken0Amount = MAXIMUM_SUPPLY - 2;
```

Comment:
If token0 gets the full supply without leaving two behind, the input transfer before the swap is going to fail. The input amount must be at least two as removing fees from one (or less) drops the output amount below one which fails the swap.